llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp

```diff
@@ -1456,6 +1456,24 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
       Opcode != Instruction::InsertElement)
     return BaseT::getVectorInstrCost(Opcode, Val, CostKind, Index, Op0, Op1);
 
+  // Extract/InsertElement with non-constant index is very costly without
+  // V instructions; estimate cost of loads/stores sequence via the stack:
+  // ExtractElement cost: store vector to stack, load scalar;
+  // InsertElement cost: store vector to stack, store scalar, load vector.
+  if (Index == -1U && !ST->hasVInstructions()) {
+    auto *VecTy = cast<VectorType>(Val);
+    auto *ElemTy = VecTy->getElementType();
+    auto NumElems = VecTy->getElementCount().getKnownMinValue();
```

Comment:
This will produce a significant under-estimate on cost for scalable vectors as the number elements will be very low.  See getEstimatedVLFor.

Reply:
Actually, I've realized that scalable vectors types can't be lowered with scalars (backend fails with "Don't know how to legalize this scalable vector type" error), and currently the reported cost of these operations is Invalid. So probably it's better to return Invalid cost instead of numeric value to show that it can't be lowered (added some test cases with vscale vectors)

Reply:
That test crashes in argument lowering, not insertelement handling.  However, you point stands, here's an alternative example: https://godbolt.org/z/13v643fGe

It does look like we're trying to scalarize the type rather than legalize the operation via the stack.